File: frameworks/micronaut-graphql/src/main/java/com/velocitybench/graphql/DataLoaderRegistry.java
package com.velocitybench.graphql;

import com.velocitybench.model.Comment;
import com.velocitybench.model.Post;
import com.velocitybench.model.User;
import com.velocitybench.repository.CommentRepository;
import com.velocitybench.repository.PostRepository;
import com.velocitybench.repository.UserRepository;
import jakarta.inject.Singleton;
import org.dataloader.BatchLoader;
import org.dataloader.DataLoader;
import org.dataloader.DataLoaderOptions;

import java.util.*;
import java.util.concurrent.CompletableFuture;

@Singleton
public class DataLoaderRegistry {

    private final UserRepository userRepository;
    private final PostRepository postRepository;
    private final CommentRepository commentRepository;

    public DataLoaderRegistry(UserRepository userRepository, PostRepository postRepository, CommentRepository commentRepository) {
        this.userRepository = userRepository;
        this.postRepository = postRepository;
        this.commentRepository = commentRepository;
    }

    public org.dataloader.DataLoaderRegistry create() {
        org.dataloader.DataLoaderRegistry registry = new org.dataloader.DataLoaderRegistry();

        registry.register("userLoader", createUserLoader());
        registry.register("postLoader", createPostLoader());
        registry.register("postsByAuthorLoader", createPostsByAuthorLoader());
        registry.register("commentsByPostLoader", createCommentsByPostLoader());

        return registry;
    }

    private DataLoader<Integer, User> createUserLoader() {
        BatchLoader<Integer, User> batchLoader = keys -> CompletableFuture.supplyAsync(() -> {
            Map<Integer, User> users = userRepository.findByPks(new HashSet<>(keys));
            return keys.stream().map(users::get).toList();
        });
        return DataLoader.newDataLoader(batchLoader, DataLoaderOptions.newOptions().setCachingEnabled(true));
    }

    private DataLoader<Integer, Post> createPostLoader() {
        BatchLoader<Integer, Post> batchLoader = keys -> CompletableFuture.supplyAsync(() -> {
            Map<Integer, Post> posts = postRepository.findByPks(new HashSet<>(keys));
            return keys.stream().map(posts::get).toList();
        });
        return DataLoader.newDataLoader(batchLoader, DataLoaderOptions.newOptions().setCachingEnabled(true));
    }

    private DataLoader<Integer, List<Post>> createPostsByAuthorLoader() {
        BatchLoader<Integer, List<Post>> batchLoader = keys -> CompletableFuture.supplyAsync(() -> {
            Map<Integer, List<Post>> postsByAuthor = postRepository.findByAuthorPks(new HashSet<>(keys), 50);
            return keys.stream().map(k -> postsByAuthor.getOrDefault(k, Collections.emptyList())).toList();
        });
        return DataLoader.newDataLoader(batchLoader, DataLoaderOptions.newOptions().setCachingEnabled(true));
    }

    private DataLoader<Integer, List<Comment>> createCommentsByPostLoader() {
        BatchLoader<Integer, List<Comment>> batchLoader = keys -> CompletableFuture.supplyAsync(() -> {
            Map<Integer, List<Comment>> commentsByPost = commentRepository.findByPostPks(new HashSet<>(keys), 50);
            return keys.stream().map(k -> commentsByPost.getOrDefault(k, Collections.emptyList())).toList();
        });
        return DataLoader.newDataLoader(batchLoader, DataLoaderOptions.newOptions().setCachingEnabled(true));
    }
}

File: frameworks/webonyx-graphql-php/src/DataLoader/DataLoaderRegistry.php
<?php
declare(strict_types=1);

namespace VelocityBench\DataLoader;

use Overblog\DataLoader\DataLoader;
use Overblog\PromiseAdapter\Adapter\WebonyxGraphQLSyncPromiseAdapter;
use VelocityBench\Model\User;
use VelocityBench\Model\Post;
use VelocityBench\Model\Comment;

class DataLoaderRegistry
{
    private WebonyxGraphQLSyncPromiseAdapter $promiseAdapter;
    private ?DataLoader $userLoader = null;
    private ?DataLoader $postLoader = null;
    private ?DataLoader $postsByAuthorLoader = null;
    private ?DataLoader $commentsByPostLoader = null;

    public function __construct()
    {
        $this->promiseAdapter = new WebonyxGraphQLSyncPromiseAdapter();
    }

    public function getUserLoader(): DataLoader
    {
        if ($this->userLoader === null) {
            $adapter = $this->promiseAdapter;
            $this->userLoader = new DataLoader(
                function (array $keys) use ($adapter) {
                    $users = User::findByPks($keys);
                    $result = array_map(fn($key) => $users[$key] ?? null, $keys);
                    return $adapter->createFulfilled($result);
                },
                $adapter
            );
        }
        return $this->userLoader;
    }

    public function getPostLoader(): DataLoader
    {
        if ($this->postLoader === null) {
            $adapter = $this->promiseAdapter;
            $this->postLoader = new DataLoader(
                function (array $keys) use ($adapter) {
                    $posts = Post::findByPks($keys);
                    $result = array_map(fn($key) => $posts[$key] ?? null, $keys);
                    return $adapter->createFulfilled($result);
                },
                $adapter
            );
        }
        return $this->postLoader;
    }

    public function getPostsByAuthorLoader(int $limit = 50): DataLoader
    {
        if ($this->postsByAuthorLoader === null) {
            $adapter = $this->promiseAdapter;
            $this->postsByAuthorLoader = new DataLoader(
                function (array $keys) use ($adapter, $limit) {
                    $postsByAuthor = Post::findByAuthorPks($keys, $limit);
                    $result = array_map(fn($key) => $postsByAuthor[$key] ?? [], $keys);
                    return $adapter->createFulfilled($result);
                },
                $adapter
            );
        }
        return $this->postsByAuthorLoader;
    }

    public function getCommentsByPostLoader(int $limit = 50): DataLoader
    {
        if ($this->commentsByPostLoader === null) {
            $adapter = $this->promiseAdapter;
            $this->commentsByPostLoader = new DataLoader(
                function (array $keys) use ($adapter, $limit) {
                    $commentsByPost = Comment::findByPostPks($keys, $limit);
                    $result = array_map(fn($key) => $commentsByPost[$key] ?? [], $keys);
                    return $adapter->createFulfilled($result);
                },
                $adapter
            );
        }
        return $this->commentsByPostLoader;
    }
}
